### DISCUSSION TITLE
Can Langium compile recursive constructs to LL(*)?

### DISCUSSION THREAD

**Discussion by trusktr** [eclipse-langium/langium#1033] [Ideas]
Would it be possible for Langium to take something like

```js
Expr: Expr '+' Expr
// ...
```

and compile it to the left-factored form for Chevrotain, so that users don't have to do it? Just curios. No idea how difficult that is.

**Comment by msujew** [comment]
Hey @trusktr,

yeah, that's pretty difficult. ANTLR4 is AFAIK the only LL-like parser with this functionality, and even they can only do it for direct left recursion. In theory that's a topic for [this repo](https://github.com/langium/chevrotain-allstar), which implements our lookahead algorithm.

Anyway, Langium isn't currently able to deal with this, which is the main reason for grammar constructs like assigned actions.

**Reply by spoenemann** [reply]
The idea has been around for a while though. One problem is that the direct recursive form does not specify precedence of multiple operators, but build the expression straight from left to right. We'd need a special syntax to denote precedence and associativity.
